### DISCUSSION TITLE
Proposal: allow typeof(this) to reference the current class

### DISCUSSION THREAD

**Discussion by LWChris** [dotnet/csharplang#8410] [Language Ideas]
For initialization of a `log4net` logger, you have two suboptimal options:

Option 1
---------
```cs
private static readonly ILog _log = LogManager.GetLogger(typeof(MyClass));
```
This works well, but you need to remember to adjust `MyClass` every time you copy that line (which happens a lot).

Option 2
---------
```cs
private static readonly ILog _log = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod()?.DeclaringType);
```

This can be copy pasted, but it's quite convoluted and it uses reflection to do what could be known at compile time.

---

Allowing to write 

```cs
private static readonly ILog _log = LogManager.GetLogger(typeof(this));
```

would mean the compiler would use the innermost class that the code is currently wrapped in. I know that `this` doesn't make sense in a static context, but allowing that keyword to appear within a `typeof()` expression, reading as "type of this" should be clear enough. It's basically the rough equivalent of VB.NET's `MyClass` keyword.

**Comment by CyrusNajmabadi** [comment]
Wouldn't `typeof(class)` make more sense?

**Reply by colejohnson66** [reply]
And what if it's a struct or record? `typeof(struct)` and `typeof(record)` would be fine with me, but `typeof(this)`, while a bit illogical, is a singular form. Not to mention what a `record struct` would be - `typeof(record struct)`?

**Comment by 333fred** [comment]
> it uses reflection to do what could be known at compile time.

My problem with this proposal is that the behavior I actually expect _requires_ reflection, or a hidden type parameter. I expect this sample to print `Derived`:

```cs
var d = new Derived();
d.M();

public class Base
{
    public void M() => Console.WriteLine(typeof(this).Name);
}

public class Derived : Base {}
```

**Reply by BlinD-HuNTeR** [reply]
You should expect "Derived" if the code was `this.GetType().Name` instead.  The proposed `typeof(this)` is supposed to mean something different, otherwise, it would be pointless to have another syntax to do the same as `GetType`.

Nevertheless, I still think the OP's problem should be addressed by source generators.

**Reply by 333fred** [reply]
Language design is rarely that absolute on what people would expect, or what would be a good idea. My initial reaction on seeing this was exactly what I wrote :slightly_smiling_face:.

**Reply by CyrusNajmabadi** [reply]
Mine too, fwiw.  Hence using 'class', which I felt properly distanced itself from 'this=instance' semantics.

**Reply by TahirAhmadov** [reply]
To me, the `typeof` is such a strong indication that this is compile-time type and not runtime instance, that I would never think `this` indicates it's runtime somehow.

**Reply by theunrepentantgeek** [reply]
Is there an existing proposal to allow generic parameters to use `class` to reference the current class?

Because it seems to me that it would be even clearer if the original suggestion was 

``` csharp
private static readonly ILog _log = LogManager.GetLogger<MyClass>();
```

with a potential rewrite to:

``` csharp
private static readonly ILog _log = LogManager.GetLogger<class>();
```

Actually, I just realized: The implentation of `LogManager.GetLogger<T>();` could include (for DEBUG builds only) a check using `MethodBase.GetCurrentMethod()` to verify that `T` is in the current inheritance hierarchy. This doesn't address the need for the code to be updated after copy/paste - but it does prevent the error from going undetected, and without any performance penalty at runtime.

**Reply by LWChris** [reply]
> Using 'class' felt properly distanced from 'this=instance' semantics.

At first I agreed when reading that.

But then I thought - _if_ I expected "Derived" when writing `typeof(this)`, would I expect "Base" when writing `typeof(class)`? And I felt like the answer would still be "no".

The quickest description of `typeof(...)` that came to my mind was "static `.GetType()`". But if I was expecting "Derived", that would imply I was oblivious to what "static" means. And that would probably be the case for `this` and `class` alike. Just a feeling though.

**Comment by BlinD-HuNTeR** [comment]
Why hasn't anyone written a source generator for this yet? Copy-pasting is a thing of the past now.

**Reply by LWChris** [reply]
I don't know a lot about source generators yet, but would that mean every class that needs a logger has to be made `partial` with a `static partial ILogger GetLogger()` method?

**Reply by BlinD-HuNTeR** [reply]
> I don't know a lot about source generators yet, but would that mean every class that needs a logger has to be made `partial` with a `static partial ILogger GetLogger()` method?

For the `partial` part, yes. But you don't need to have a `GetLogger` method, you could have the source generator produce the `_log` field directly.

**Comment by TahirAhmadov** [comment]
https://github.com/dotnet/csharplang/discussions/5261

**Reply by LWChris** [reply]
Bah. [Search](https://github.com/dotnet/csharplang/discussions?discussions_q=is:open%20%22typeof%22%20%22this%22) doesn't show that one on the first 3 pages... Of course, it is the first entry on page 4. 🙄
